Update _P076_HLW8012.ino / Add support for BL0937 to plugin _076_HLW8012

CaptainRoot opened the pull request:
Add support for BL0937  to plugin _076_HLW8012
Add possibility to set interrupt trigger for cf and cf1
Add possibility to set current reading state (low/high)

CaptainRoot committed:
Update _P076_HLW8012.ino

Add support for BL0937  to plugin _076_HLW8012
-Add possibility to set interrupt tigger for cf and cf1
Add possibility to set current reading state (low/high)

TD-er commented:
In what devices is the BL0937 used?
Maybe we can also add a selector to choose some predefined config and as last option "custom".

CaptainRoot commented:
> In what devices is the BL0937 used?
> Maybe we can also add a selector to choose some predefined config and as last option "custom".

BL0937 is used in:
Aplic WDP303075, BlitzWolf SHP, Gosund SP1 v23 and some Teckin (based on Sonoff-Tasmota)
Predefined configs seams to be a good idea. i will try to add predefined configs.

CaptainRoot committed:
add predefind pin settings for some energy meter sockets to HLW8012 / BL0937

CaptainRoot committed:
Revert changes to platformio.ini

CaptainRoot committed:
Update platformio.ini

CaptainRoot commented:
I added some predefined settings

TD-er commented on platformio.ini:
```diff
@@ -14,7 +14,7 @@
 ;env_default = dev_ESP8266_4096
 ;env_default = normal_ESP8266_1024
-;env_default = normal_ESP8266_4096
+env_default = normal_ESP8266_4096
```
Not sure what it was, but a (long) time ago, we put those default settings in comments.
I cannot remember the reason why.

CaptainRoot replied:
I will revert that change

TD-er commented on src/_P076_HLW8012.ino:
```diff
@@ -55,6 +55,39 @@ unsigned int p076_hvoltage = 0;
+typedef struct p076_DevicePinSettings P076_DevicePinSettings;
+
+const int p076_PinSettingsCount = 10;
+P076_DevicePinSettings p076_PredefinedPinSettings[p076_PinSettingsCount] =
```
Maybe this can be stored as `PROGMEM` ?
Like it is defined now, it takes quite a lot of memory from the sys stack (NonOS SDK stack).

CaptainRoot replied:
Moved to PROGMEM, now

TD-er commented on src/_P076_HLW8012.ino:
```diff
@@ -104,11 +137,41 @@ boolean Plugin_076(byte function, struct EventStruct *event, String &string) {
+    predefinedId[6] = p076_PredefinedPinSettings[6].Id;
+    predefinedId[7] = p076_PredefinedPinSettings[7].Id;
+    predefinedId[8] = p076_PredefinedPinSettings[8].Id;
+    predefinedId[9] = p076_PredefinedPinSettings[9].Id;
```
The `Id` is just sequential, so why not use a loop to fill it?

CaptainRoot replied:
good idea, will be fixed

TD-er commented on src/_P076_HLW8012.ino:
```diff
@@ -104,11 +137,41 @@ boolean Plugin_076(byte function, struct EventStruct *event, String &string) {
+    predefinedNames[6] = p076_PredefinedPinSettings[6].Device_Name;
+    predefinedNames[7] = p076_PredefinedPinSettings[7].Device_Name;
+    predefinedNames[8] = p076_PredefinedPinSettings[8].Device_Name;
+    predefinedNames[9] = p076_PredefinedPinSettings[9].Device_Name;
```
Why not use a loop to fill the `predefinedNames` ?
Both indices are the same and sequential.

CaptainRoot replied:
good idea, will be fixed

TD-er commented on src/_P076_HLW8012.ino:
```diff
@@ -165,25 +240,56 @@ boolean Plugin_076(byte function, struct EventStruct *event, String &string) {
+    }
+
+    //Set Pin settings
+    byte selectedDevice = getFormItemInt(F("p076_predefined_settings"));
```
Try to shorten the internal identifiers.
The compiled binary is getting too big, so we must try to de-duplicate code as much as possible and since those internal identifiers are just used to recognize returned values from the web interface, they can be made shorter. No human is reading them, only when debugging.

CaptainRoot replied:
Fixed, shorten to p076_preDefDevSel

TD-er commented on src/_P076_HLW8012.ino:
```diff
@@ -165,25 +240,56 @@ boolean Plugin_076(byte function, struct EventStruct *event, String &string) {
+    }
+
+    if (PLUGIN_076_DEBUG) {
+      String log = F("HLW8012: PIN Settings ");
```
The plugin will now support not only the HWL8012, so maybe just mention the plugin-ID in log sentences? (there are more of these)

CaptainRoot replied:
Fixed

CaptainRoot committed:
movepredefind pin settings to PROGMEM

move predefind pin settings for some energy meter sockets to HLW8012 /BL0937 to PROGMEM

CaptainRoot committed:
Merge branch 'mega' of https://github.com/letscontrolit/ESPEasy into mega

TD-er commented on src/_P076_HLW8012.ino:
```diff
@@ -67,27 +56,40 @@ struct p076_DevicePinSettings {
+   { 9, string_Gosund,     12,      4,     5,       LOW,    FALLING, CHANGE}
 };
 
+P076_PredefinedDevice p076_SelectedPredefinedDevice;
```
Please either declare this on the stack, inside the function where you use it, or make it an allocation on the heap when needed (using a pointer and new... )
Like this, it is created and always consuming memory, even if the plugin is not active.

TD-er commented:
I will merge it now and address the last points I made.
Thanks for these fixes.
